# Getting started with R

**By [Christine Zhang](https://twitter.com/christinezhang) (Knight-Mozilla / Los Angeles Times) & [Ryan Menezes](https://twitter.com/ryanvmenezes) (Los Angeles Times)**

*IRE Conference -- New Orleans, LA*
 
June 18, 2016  

This workshop is a basic introduction to R, a free, open-source software for data analysis and statistics. 

R is a powerful tool that can help you quickly and effectively answer questions using data.

Take our host city, New Orleans, for example. Hurricane Katrina was a devastating natural disaster that substantially affected the population of New Orleans. The hurricane took place in August 2005, which coincidentally falls between the U.S. Census full population counts in 2000 and 2010.

In this session, we will use the "Demographic Profile” -- a large summary file with many different demographic variables downloaded from the [U.S. Census Bureau website](http://factfinder.census.gov/faces/nav/jsf/pages/index.xhtml) -- from 2000 and 2010, for all census tracts in the state of Louisiana.

**In this session, we will:**

* Load the 2000 data in R
* Select a few variables pertaining to population counts and housing occupancy
* Clean the data
* Focus on the Orleans Parish in particular, which represents the city of New Orleans, and tally the population
* Perform the same steps on the 2010 data
* Merge the two sets
* Write the merged data out to a CSV file for our next class, [More with R](More%20with%20R.ipynb), where we'll do some more in depth analysis

Basic analysis techniques like the ones you will learn in this class can help you write data-driven stories, like [this one written by The Times-Picayune](http://www.nola.com/politics/index.ssf/2011/02/new_orleans_officials_2010_pop.html) shortly after the census released its 2010 tally.

The story begins:

> Five years after Hurricane Katrina emptied New Orleans and prompted the largest mass migration in modern American history, **the 2010 Census counted 343,829 people living in the still-recovering city, a 29 percent drop since the last head count a decade ago**, according to data released today.

Using the data we have, we will attempt to replicate the calculations in that lede.

The following code and annotations were written in a Jupyter notebook. The code is best run in RStudio using R version 3.3.0

### Loading data

We'll start by loading in the 2000 data, which is stored in a CSV (comma-separated values) file. CSVs are plain-text files of data where commas separate the columns within a line. It is sometimes preferable to work with CSVs as opposed to files of a proprietary format, such as Microsoft Excel files, but the Census Bureau readily makes data available in both formats.

Let's run R's `read.csv` command and save the data to an object called `census2000`. Here, we are using assignment with `<-`, which tells R to run the right side and assign the result to the object named on the left. 

In [1]:
census2000 <- read.csv('2000_census_demographic_profile.csv')

Now that this ran without incident, let's inspect the first few rows using `head`, which by default prints out the first six rows of a data frame (R's internal term for a spreadsheet):

In [2]:
head(census2000)

GEO.id     GEO.id2                           GEO.display.label
1                   Id         Id2                                   Geography
2 1400000US22001960100 22001960100 Census Tract 9601, Acadia Parish, Louisiana
3 1400000US22001960200 22001960200 Census Tract 9602, Acadia Parish, Louisiana
4 1400000US22001960300 22001960300 Census Tract 9603, Acadia Parish, Louisiana
5 1400000US22001960400 22001960400 Census Tract 9604, Acadia Parish, Louisiana
6 1400000US22001960500 22001960500 Census Tract 9605, Acadia Parish, Louisiana
                 HC01_VC01                 HC02_VC01
1 Number; Total population Percent; Total population
2                    6,188                       100
3                    5,056                       100
4                    3,149                       100
5                    5,617                       100
6                    4,927                       100
                                      HC01_VC03
1 Number; Total population - SEX AND AGE - Male
2                                         2,920
3                                         2,562
4                                         1,593
5                                         2,754
6                                         2,461
                                       HC02_VC03
1 Percent; Total population - SEX AND AGE - Male
2                                             47
3                                             51
4                                             51
5                                             49
6                                             50
                                        HC01_VC04
1 Number; Total population - SEX AND AGE - Female
2                                           3,268
3                                           2,494
4                                           1,556
5                                           2,863
6                                           2,466
                                         HC02_VC04
1 Percent; Total population - SEX AND AGE - Female
2                                               53
3                                               49
4                                               49
5                                               51
6                                               50
                                               HC01_VC05
1 Number; Total population - SEX AND AGE - Under 5 years
2                                                    462
3                                                    346
4                                                    209
5                                                    429
6                                                    400
                                                HC02_VC05
1 Percent; Total population - SEX AND AGE - Under 5 years
2                                                       8
3                                                       7
4                                                       7
5                                                       8
6                                                       8
                                              HC01_VC06
1 Number; Total population - SEX AND AGE - 5 to 9 years
2                                                   502
3                                                   416
4                                                   251
5                                                   406
6                                                   438
                                               HC02_VC06
1 Percent; Total population - SEX AND AGE - 5 to 9 years
2                                                      8
3                                                      8
4                                                      8
5                                                      7
6                                                      9
                                                HC01_VC07
1 Number; Total population - SEX AND AGE - 10 to 14 

Upon inspection, we can see that the file came with two header rows. R, by default, takes the first row of a CSV to be the header. We clearly do not need the first one so we can rerun the `read.csv` command and tell it so:

In [3]:
census2000 <- read.csv('2000_census_demographic_profile.csv', skip = 1)

In [4]:
head(census2000)

Id         Id2                                   Geography
1 1400000US22001960100 22001960100 Census Tract 9601, Acadia Parish, Louisiana
2 1400000US22001960200 22001960200 Census Tract 9602, Acadia Parish, Louisiana
3 1400000US22001960300 22001960300 Census Tract 9603, Acadia Parish, Louisiana
4 1400000US22001960400 22001960400 Census Tract 9604, Acadia Parish, Louisiana
5 1400000US22001960500 22001960500 Census Tract 9605, Acadia Parish, Louisiana
6 1400000US22001960600 22001960600 Census Tract 9606, Acadia Parish, Louisiana
  Number..Total.population Percent..Total.population
1                    6,188                       100
2                    5,056                       100
3                    3,149                       100
4                    5,617                       100
5                    4,927                       100
6                    5,654                       100
  Number..Total.population...SEX.AND.AGE...Male
1                                         2,920
2                                         2,562
3                                         1,593
4                                         2,754
5                                         2,461
6                                         2,647
  Percent..Total.population...SEX.AND.AGE...Male
1                                             47
2                                             51
3                                             51
4                                             49
5                                             50
6                                             47
  Number..Total.population...SEX.AND.AGE...Female
1                                           3,268
2                                           2,494
3                                           1,556
4                                           2,863
5                                           2,466
6                                           3,007
  Percent..Total.population...SEX.AND.AGE...Female
1                                               53
2                                               49
3                                               49
4                                               51
5                                               50
6                                               53
  Number..Total.population...SEX.AND.AGE...Under.5.years
1                                                    462
2                                                    346
3                                                    209
4                                                    429
5                                                    400
6                                                    464
  Percent..Total.population...SEX.AND.AGE...Under.5.years
1                                                       8
2                                                       7
3                                                       7
4                                                       8
5                                                       8
6                                                       8
  Number..Total.population...SEX.AND.AGE...5.to.9.years
1                                                   502
2                                                   416
3                                                   251
4                                                   406
5                                                   438
6                                                   471
  Percent..Total.population...SEX.AND.AGE...5.to.9.years
1                                                      8
2                                                      8
3                                                      8
4                                                      7
5                                                      9
6                                                      8
  Number..Total.population...SEX.AND.AGE...10.to.14.years
1                                                     54

### Cleaning the data

Visually, we can see that this data set is very wide. In fact, there are 195 columns.

Spaces are not allowed in R column names. That's why they've been automatically converted to periods, as in `Number..Total.population`.

Let's keep a handful of these:

* `Id2`: This is what the census bureau calls a FIPS code. It is a unique numerical identifier for all census tracts. This will be important when we join our two datasets together.  
 
 
* `Geography`: This is a text description of the tract, with the parish name.  
 
 
* `Number..Total.population`: The total population of the tract.  
 
 
* `Number..HOUSING.OCCUPANCY...Total.housing.units`, `Number..HOUSING.OCCUPANCY...Total.housing.units...Occupied.housing.units`, and `Number..HOUSING.OCCUPANCY...Total.housing.units...Vacant.housing.units`: The total, occupied and vacant housing units.

To help us trim the data set to just these six columns, we are going to import a package. There are thousands of packages for R created by the open-source community, which help improve on what is included in R by default.

The one we will use here is called dplyr.

In [5]:
## if dplyr was not installed we would have to run this
# install.packages('dplyr')

## to import the package and all of its functions 
library('dplyr')

Warning message:
: package ‘dplyr’ was built under R version 3.2.4
Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



From dplyr, we will use the `select` function to trim the data set and save it to a new variable called census2000.trimmed:

In [6]:
census2000.trimmed <- select(
    census2000, # name of the data frame
    # list of all the six column names we want to keep
    Id2, 
    Geography, 
    Number..Total.population, 
    Number..HOUSING.OCCUPANCY...Total.housing.units, 
    Number..HOUSING.OCCUPANCY...Total.housing.units...Occupied.housing.units, 
    Number..HOUSING.OCCUPANCY...Total.housing.units...Vacant.housing.units
)

head(census2000.trimmed)

Id2                                   Geography
1 22001960100 Census Tract 9601, Acadia Parish, Louisiana
2 22001960200 Census Tract 9602, Acadia Parish, Louisiana
3 22001960300 Census Tract 9603, Acadia Parish, Louisiana
4 22001960400 Census Tract 9604, Acadia Parish, Louisiana
5 22001960500 Census Tract 9605, Acadia Parish, Louisiana
6 22001960600 Census Tract 9606, Acadia Parish, Louisiana
  Number..Total.population Number..HOUSING.OCCUPANCY...Total.housing.units
1                    6,188                                           2,410
2                    5,056                                           1,909
3                    3,149                                           1,246
4                    5,617                                           2,176
5                    4,927                                           1,796
6                    5,654                                           2,292
  Number..HOUSING.OCCUPANCY...Total.housing.units...Occupied.housing.units
1                                                                    2,236
2                                                                    1,764
3                                                                    1,145
4                                                                    1,991
5                                                                    1,692
6                                                                    2,073
  Number..HOUSING.OCCUPANCY...Total.housing.units...Vacant.housing.units
1                                                                    174
2                                                                    145
3                                                                    101
4                                                                    185
5                                                                    104
6                                                                    219

This shows us that we were able to select the columns correctly. But one lingering issue is that these column names are long and unwieldy. Since we are going to be typing them often, let's rename them to shorter, more convenient versions:

In [7]:
colnames(census2000.trimmed) <- c(
    'fips.code', 'geography', 'population',
    'total.housing.units', 'occupied.housing.units', 'vacant.housing.units'
)
head(census2000.trimmed)

fips.code                                   geography population
1 22001960100 Census Tract 9601, Acadia Parish, Louisiana      6,188
2 22001960200 Census Tract 9602, Acadia Parish, Louisiana      5,056
3 22001960300 Census Tract 9603, Acadia Parish, Louisiana      3,149
4 22001960400 Census Tract 9604, Acadia Parish, Louisiana      5,617
5 22001960500 Census Tract 9605, Acadia Parish, Louisiana      4,927
6 22001960600 Census Tract 9606, Acadia Parish, Louisiana      5,654
  total.housing.units occupied.housing.units vacant.housing.units
1               2,410                  2,236                  174
2               1,909                  1,764                  145
3               1,246                  1,145                  101
4               2,176                  1,991                  185
5               1,796                  1,692                  104
6               2,292                  2,073                  219

Another helpful command to run on any data set is `str`, which gives you the structure of the variable as defined by R:

In [8]:
str(census2000.trimmed)

'data.frame':	1106 obs. of  6 variables:
 $ fips.code             : num  2.2e+10 2.2e+10 2.2e+10 2.2e+10 2.2e+10 ...
 $ geography             : Factor w/ 1106 levels "Census Tract 10.01, Lafayette Parish, Louisiana",..: 970 978 985 993 1000 1006 1011 1015 1018 1021 ...
 $ population            : Factor w/ 1019 levels "0","1","10,248",..: 859 710 350 801 692 806 647 804 711 842 ...
 $ total.housing.units   : Factor w/ 905 levels "1","10","1,002",..: 565 404 104 499 357 531 397 522 426 594 ...
 $ occupied.housing.units: Factor w/ 876 levels "0","1","1,001",..: 512 363 74 446 338 468 347 470 374 514 ...
 $ vacant.housing.units  : Factor w/ 374 levels "0","1","10","100",..: 86 54 5 98 9 134 83 86 110 196 ...


The structure tells us that this is a data frame with 1106 rows and six columns. It further tells us the type of each column. 

Notice how the FIPS code read in as a number but the other numeric columns read in as “factors”? That's R-speak for a categorical variable, and any character variables are by default set to this type. This happened because the numbers and those columns have commas. The presence of a single character within a number makes R treat the entire column as strings. This will be an issue later when we try to add two numbers together, as R doesn't know how to add two characters.

The solution: we need to remove the comma from all the strings, then recast the variable as a number.

To help with this we are going to use another package called `stringr`, and a function from within it called `str_replace`:

In [9]:
# install.packages('stringr')

library('stringr')

Warning message:
: package ‘stringr’ was built under R version 3.2.5

Let's start with the population variable. First, let's remove the comma and write the result to the original column. (The format for calling a column from a data frame in R is `df.name$column.name`)

In [10]:
census2000.trimmed$population <- str_replace(
    census2000.trimmed$population, 
    pattern = ',', 
    replacement = ''
)

Then we'll visually inspect the head:

In [11]:
head(census2000.trimmed)

fips.code                                   geography population
1 22001960100 Census Tract 9601, Acadia Parish, Louisiana       6188
2 22001960200 Census Tract 9602, Acadia Parish, Louisiana       5056
3 22001960300 Census Tract 9603, Acadia Parish, Louisiana       3149
4 22001960400 Census Tract 9604, Acadia Parish, Louisiana       5617
5 22001960500 Census Tract 9605, Acadia Parish, Louisiana       4927
6 22001960600 Census Tract 9606, Acadia Parish, Louisiana       5654
  total.housing.units occupied.housing.units vacant.housing.units
1               2,410                  2,236                  174
2               1,909                  1,764                  145
3               1,246                  1,145                  101
4               2,176                  1,991                  185
5               1,796                  1,692                  104
6               2,292                  2,073                  219

This appeared to work. But R will still think this is a character variable unless we explicitly tell it otherwise:

In [12]:
census2000.trimmed$population <- as.numeric(census2000.trimmed$population)

Running `str` will help us ensure this worked:

In [13]:
str(census2000.trimmed)

'data.frame':	1106 obs. of  6 variables:
 $ fips.code             : num  2.2e+10 2.2e+10 2.2e+10 2.2e+10 2.2e+10 ...
 $ geography             : Factor w/ 1106 levels "Census Tract 10.01, Lafayette Parish, Louisiana",..: 970 978 985 993 1000 1006 1011 1015 1018 1021 ...
 $ population            : num  6188 5056 3149 5617 4927 ...
 $ total.housing.units   : Factor w/ 905 levels "1","10","1,002",..: 565 404 104 499 357 531 397 522 426 594 ...
 $ occupied.housing.units: Factor w/ 876 levels "0","1","1,001",..: 512 363 74 446 338 468 347 470 374 514 ...
 $ vacant.housing.units  : Factor w/ 374 levels "0","1","10","100",..: 86 54 5 98 9 134 83 86 110 196 ...


For the rest of the columns we can nest the first function within the second to speed things up:

In [14]:
census2000.trimmed$total.housing.units <- as.numeric(str_replace(census2000.trimmed$total.housing.units, pattern = ',', replacement = ''))
census2000.trimmed$occupied.housing.units <- as.numeric(str_replace(census2000.trimmed$occupied.housing.units, pattern = ',', replacement = ''))
census2000.trimmed$vacant.housing.units <- as.numeric(str_replace(census2000.trimmed$vacant.housing.units, pattern = ',', replacement = ''))

In [15]:
str(census2000.trimmed)

'data.frame':	1106 obs. of  6 variables:
 $ fips.code             : num  2.2e+10 2.2e+10 2.2e+10 2.2e+10 2.2e+10 ...
 $ geography             : Factor w/ 1106 levels "Census Tract 10.01, Lafayette Parish, Louisiana",..: 970 978 985 993 1000 1006 1011 1015 1018 1021 ...
 $ population            : num  6188 5056 3149 5617 4927 ...
 $ total.housing.units   : num  2410 1909 1246 2176 1796 ...
 $ occupied.housing.units: num  2236 1764 1145 1991 1692 ...
 $ vacant.housing.units  : num  174 145 101 185 104 219 171 174 196 284 ...


By default, `head` will print the first six lines. But we can override the default to show as many as we want (we'll show 10 here):

In [16]:
head(census2000.trimmed, n = 10)

fips.code                                   geography population
1  22001960100 Census Tract 9601, Acadia Parish, Louisiana       6188
2  22001960200 Census Tract 9602, Acadia Parish, Louisiana       5056
3  22001960300 Census Tract 9603, Acadia Parish, Louisiana       3149
4  22001960400 Census Tract 9604, Acadia Parish, Louisiana       5617
5  22001960500 Census Tract 9605, Acadia Parish, Louisiana       4927
6  22001960600 Census Tract 9606, Acadia Parish, Louisiana       5654
7  22001960700 Census Tract 9607, Acadia Parish, Louisiana       4614
8  22001960800 Census Tract 9608, Acadia Parish, Louisiana       5640
9  22001960900 Census Tract 9609, Acadia Parish, Louisiana       5059
10 22001961000 Census Tract 9610, Acadia Parish, Louisiana       5965
   total.housing.units occupied.housing.units vacant.housing.units
1                 2410                   2236                  174
2                 1909                   1764                  145
3                 1246                   1145                  101
4                 2176                   1991                  185
5                 1796                   1692                  104
6                 2292                   2073                  219
7                 1894                   1723                  171
8                 2254                   2080                  174
9                 1978                   1782                  196
10                2526                   2242                  284

That worked!

But in the interest of full disclosure, you should know that we added those commas to the original CSVs from the Census Bureau to facilitate this exercise. “Commafied” numbers are one of the most frequent stumbling blocks to creating a cleaned data set.

For our last cleaning exercise, we'll work with the geography column. It has a lot of information in there, but it would be more useful if the census tract, parish name and state were separated, to help us aggregate some of these numbers.

The package tidyr has a function that helps us do just that:

In [17]:
# install.packages('tidyr')

library('tidyr')

Should you run into a function and not know what arguments it takes, running the function name, a pair of of empty parentheses afterwards, preceded by a question mark will allow you to access the documentation on that function:

In [18]:
# ?separate()

In [19]:
census2000.trimmed <- separate(
    census2000.trimmed, # name of the data frame
    geography, # column to split
    c('tract', 'parish', 'state'), # new column names
    ', ' # delimiter to split on (note the space after the comma)
)

In [20]:
head(census2000.trimmed)

fips.code             tract        parish     state population
1 22001960100 Census Tract 9601 Acadia Parish Louisiana       6188
2 22001960200 Census Tract 9602 Acadia Parish Louisiana       5056
3 22001960300 Census Tract 9603 Acadia Parish Louisiana       3149
4 22001960400 Census Tract 9604 Acadia Parish Louisiana       5617
5 22001960500 Census Tract 9605 Acadia Parish Louisiana       4927
6 22001960600 Census Tract 9606 Acadia Parish Louisiana       5654
  total.housing.units occupied.housing.units vacant.housing.units
1                2410                   2236                  174
2                1909                   1764                  145
3                1246                   1145                  101
4                2176                   1991                  185
5                1796                   1692                  104
6                2292                   2073                  219

Our data set is as cleaned up as we need it to be now.

Let's summarize it with a frequency table of the county names:

In [21]:
table(census2000.trimmed$parish)


              Acadia Parish                Allen Parish 
                         12                           5 
           Ascension Parish           Assumption Parish 
                         14                           6 
           Avoyelles Parish           Beauregard Parish 
                          9                           7 
           Bienville Parish              Bossier Parish 
                          5                          19 
               Caddo Parish            Calcasieu Parish 
                         64                          41 
            Caldwell Parish              Cameron Parish 
                          3                           2 
           Catahoula Parish            Claiborne Parish 
                          3                           5 
           Concordia Parish              De Soto Parish 
                          5                           7 
    East Baton Rouge Parish         East Carroll Parish 
                         89   

Now we need to run all of the above cleaning steps on the 2010 data:

In [22]:
census2010 <- read.csv('2010_census_demographic_profile.csv', skip = 1)

census2010.trimmed <- select(
  census2010, # name of the data frame
  # list of all the column names we want to keep
  Id2, Geography, Number..SEX.AND.AGE...Total.population, 
  Number..HOUSING.OCCUPANCY...Total.housing.units, 
  Number..HOUSING.OCCUPANCY...Total.housing.units...Occupied.housing.units, 
  Number..HOUSING.OCCUPANCY...Total.housing.units...Vacant.housing.units
)

colnames(census2010.trimmed) <- c('fips.code', 'census.tract', 'population', 
                               'total.housing.units', 'occupied.housing.units', 'vacant.housing.units')

census2010.trimmed$population <- as.numeric(str_replace(census2010.trimmed$population, pattern = ',', replacement = ''))
census2010.trimmed$total.housing.units <- as.numeric(str_replace(census2010.trimmed$total.housing.units, pattern = ',', replacement = ''))
census2010.trimmed$occupied.housing.units <- as.numeric(str_replace(census2010.trimmed$occupied.housing.units, pattern = ',', replacement = ''))
census2010.trimmed$vacant.housing.units <- as.numeric(str_replace(census2010.trimmed$vacant.housing.units, pattern = ',', replacement = ''))

census2010.trimmed <- separate(census2010.trimmed, census.tract, c('tract', 'parish', 'state'), ', ')

orleans2010 <- filter(census2010.trimmed, parish == 'Orleans Parish')

### Merging

Now that we've cleaned both of our data files, let's merge the 2000 and 2010 data. Merging allows you to link two data sets on values common to both. It is a powerful operation that cannot be easily done in a program like Excel with such versatility.

In this case, we know that the FIPS code and the character names for most of the tracts should be consistent across the 10-year period.

However, census tracts are added, deleted, split and joined over the course of 10 years. We will make sure to keep all entries in both years. This is what is referred to as a "full outer join.” If we were to only keep all rows that were common to both data frames (R’s default behavior) we would lose some data.

In [23]:
census.comparison <- merge(
    census2000.trimmed, # first data frame
    census2010.trimmed, # second data frame
    by = c('fips.code', 'tract', 'parish', 'state'), # keys to use for join
    suffixes = c('.00', '.10'), # suffixes to append to new columns
    all = TRUE # specifying to keep all data from both data frames
)

Let's inspect a portion of the data frame where there are full matches and partial matches:

In [24]:
census.comparison[65:69, ]

fips.code               tract         parish     state population.00
65 22015010801 Census Tract 108.01 Bossier Parish Louisiana          3159
66 22015010803 Census Tract 108.03 Bossier Parish Louisiana          5362
67 22015010804 Census Tract 108.04 Bossier Parish Louisiana          6101
68 22015010805 Census Tract 108.05 Bossier Parish Louisiana            NA
69 22015010806 Census Tract 108.06 Bossier Parish Louisiana            NA
   total.housing.units.00 occupied.housing.units.00 vacant.housing.units.00
65                   1527                      1351                     176
66                   2250                      2120                     130
67                   2383                      2282                     101
68                     NA                        NA                      NA
69                     NA                        NA                      NA
   population.10 total.housing.units.10 occupied.housing.units.10
65          3359                   1415                      1284
66            NA                     NA                        NA
67          7278                   2943                      2815
68          3238                   1585                      1425
69          4086                   1671                      1610
   vacant.housing.units.10
65                     131
66                      NA
67                     128
68                     160
69                      61

### Saving your work

Saving your intermediate work to a file is often good practice, so we will write the results of our merge to a CSV (you can do this with any data frame you create in R).

In [25]:
write.csv(census.comparison, 'census_comparison.csv', row.names = FALSE)

### Calculating population changes in New Orleans

Let's filter our merged data frame down to just Orleans Parish. The Orleans Parish and the city of New Orleans are “coterminous” (that is, they share the same boundaries), so this will isolate only the census tracts of the city.

In [26]:
# note the use of "==" since we are expressing a criterion
orleans <- filter(census.comparison, parish == 'Orleans Parish') 
head(orleans)

fips.code             tract         parish     state population.00
1 2.2071e+10    Census Tract 1 Orleans Parish Louisiana          2381
2 2.2071e+10    Census Tract 2 Orleans Parish Louisiana          1347
3 2.2071e+10    Census Tract 3 Orleans Parish Louisiana          1468
4 2.2071e+10    Census Tract 4 Orleans Parish Louisiana          2564
5 2.2071e+10 Census Tract 6.01 Orleans Parish Louisiana          2034
6 2.2071e+10 Census Tract 6.02 Orleans Parish Louisiana          2957
  total.housing.units.00 occupied.housing.units.00 vacant.housing.units.00
1                   1408                      1145                     263
2                    691                       496                     195
3                    719                       559                     160
4                   1034                       873                     161
5                    704                       506                     198
6                   1106                      1011                      95
  population.10 total.housing.units.10 occupied.housing.units.10
1          2455                   1513                      1229
2          1197                    738                       496
3          1231                    641                       467
4          2328                   1137                       911
5           849                    328                       269
6          2534                   1108                       923
  vacant.housing.units.10
1                     284
2                     242
3                     174
4                     226
5                      59
6                     185

Now we can do some quick calculations with our new merged data frame for New Orleans.

First question: **What was the population of New Orleans in 2000?**

That requires summing up the 2000 population column like so:

In [27]:
sum(orleans$population.00)

[1] NA

Why didn't this work? Let's inspect the population.00 variable using `summary`:

In [28]:
summary(orleans$population.00)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
     52    1711    2274    2678    3141    9931      30 

This reveals that there are 30 census tracts that have NA, or missing, values for population.00. By default, R does not compute a sum of a column if there are missing values. We'll have to tell it to ignore these missing values by specifying `na.rm = TRUE`:

In [29]:
sum(orleans$population.00, na.rm = TRUE)

[1] 484674

Second question: **What was the population of New Orleans in 2010?**

In [30]:
sum(orleans$population.10, na.rm = TRUE)

[1] 343829

This matches the story exactly.

Last question: **What was the percent change in New Orleans population between 2000 and 2010?**

To do this, we'll first save each population calculation to new objects. Then we'll create another object to store the percent change.

In [31]:
nola2000pop <- sum(orleans$population.00, na.rm = TRUE)
nola2010pop <- sum(orleans$population.10, na.rm = TRUE)

perc.change.nola <- (nola2010pop - nola2000pop)/nola2000pop * 100

In [32]:
print(paste('The percent change in New Orleans population since 2000 is ', round(perc.change.nola), '%', sep =''))

[1] "The percent change in New Orleans population since 2000 is -29%"


Again, we see that this matches the 29% drop cited by [The Times-Picayune article]( http://www.nola.com/politics/index.ssf/2011/02/new_orleans_officials_2010_pop.html). Yay!

This concludes our workshop, Getting started with R. 

We'll use the merged data CSV we saved above to do further analysis in our next workshop, [More with R](http://www.ire.org/events-and-training/event/2199/2638/). Take a sneak peak at the notebook [here](More%20with%20R.ipynb).

Any questions?

* christine.zhang@latimes.com or [@christinezhang](https://twitter.com/christinezhang) on Twitter
* ryan.menezes@latimes.com or [@ryanvmenezes](https://twitter.com/ryanvmenezes) on Twitter